### ISSUE TITLE
Idea to run `Revise.track()` automatically on an entire project

### ISSUE THREAD

**Issue by xximranxx** [timholy/Revise.jl#78]
Hi

I think it would be ideal if we could do `Revise.track(filename)` and this would also `Revise.track()` on all the `import`'s, `using`'s

One could do the following:
A file contains a list of all the file paths relevant to the project at hand
`Revise` goes through each, notes module name per file path
`Revise` goes through each, tries `track` on the file path
 If `track` fails, find `import` and `using` statements, try tracking these first given the file (file path and module name) -> this uses a recursive function
 Else move to next file path

Basically creating a tree

Don't forget to `error` if circle found (and tree cannot be built)

This way, one defines, in any order, the list of files used by a project, runs a `Revise` command once and everything is loaded and tracked

(I hope this is the correct place to write this)

**Comment by timholy** [comment]
Anything that you can access via `using` or `import` already gets tracked automatically. You only need `Revise.track(filename)` for "scripts" that you're running via `include(filename)`.

If you're running Julia 0.6, the version of Revise for Julia 0.7 does a much better job (hopefully, a perfect job) of identifying dependencies. Unfortunately that's just not possible on 0.6.

**Comment by xximranxx** [comment]
But does one not always have to first `include` all the files of the project
Only `import` or `using` is not enough, because the file path is unknown
During while using `Revise.track()` to do the `include`, the order of `track` statements if important
Or am I wrong?

**Comment by timholy** [comment]
`using` is typically enough, especially for precompiled packages, because the paths get stored in the `*.ji` file. On 0.7, all modules store their dependencies.

Try the demo with `Example.jl` in the README; you don't need to `include` any file.

**Comment by xximranxx** [comment]
But is this not only true for packages?
I wondering about the dependency amongst my own `module`'s
Those have to be first included in the correct order, I think

Or is the solution that I should be always making my own, local packages instead of just loose modules?

**Comment by timholy** [comment]
Can you please provide a demonstration of something that isn't working for you? I'm having a hard time guessing the precise nature of your concern.

**Comment by xximranxx** [comment]
A.jl:
```
module A
f() = 1
end
```

B.jl:
```
module B
import A
g() = A.f()
end
```

To use `Revise` on these two modules, I need to do 
```
Revise.track("A.jl")
Revise.track("B.jl")
```
IN THIS ORDER

If I did `Revise.track("B.jl")` first, it would fail, as it would not know about `module A`

Is this not something, everyone should be facing when working on larger projects?
I must be overlooking a trivial solution, because I don't see anyone writing about it and discourse topics that I started on this (more than once) are ignored :/

**Comment by timholy** [comment]
This is on Julia 0.7, but I think it should also work on 0.6:
```julia
shell> mkdir /tmp/mycode

julia> push!(LOAD_PATH, "/tmp/mycode")
5-element Array{Any,1}:
 Pkg.dir                                                                                                                                                               
 Base.AbstractEnv[Base.CurrentEnv(), Base.NamedEnv("v0.7.0"), Base.NamedEnv("v0.7"), Base.NamedEnv("v0"), Base.NamedEnv("default"), Base.NamedEnv("v0.7", create=true)]
 "/home/tim/src/julia-1.0/usr/local/share/julia/site/v0.7"                                                                                                             
 "/home/tim/src/julia-1.0/usr/share/julia/site/v0.7"                                                                                                                   
 "/tmp/mycode"                                                                                                                                                         

# Create the A.jl and B.jl files...

# ...and then test them:
julia> using A, B
[ Info: Recompiling stale cache file /home/tim/.julia/lib/v0.7/A.ji for module A
[ Info: Recompiling stale cache file /home/tim/.julia/lib/v0.7/B.ji for module B

julia> B.g()
1

# Now edit the definition of `g` so that it just returns 7. Then...

julia> B.g()
┌ Warning: `endof(a)` is deprecated, use `lastindex(a)` instead.
│   caller = parse_source!(::Dict{Module,Revise.ExprsSigs}, ::String, ::Symbol, ::Int64, ::Module) at parsing.jl:53
└ @ Revise parsing.jl:53
7
```

**Comment by xximranxx** [comment]
Aaahh - So this is the magic line that I was missing
`push!(LOAD_PATH, code_path)`

After editing the LOAD_PATH, I can do `using A,B` without `include`

The redefinition of `A.f()` also automatically changes the value of `B.g()`
BUT only in a normal REPL, not in the (new) Juno REPL
In the Juno REPL, `B.g()` does not reflect the changed code, not if changed in B.jl nor in A.jl

Maybe 0.7 works better - I will try it (currently using julia 0.6.2)

**Comment by timholy** [comment]
For anything Juno-specific, perhaps report it there and CC me?

I'll close this since I think there are no changes to Revise needed?

**Comment by xximranxx** [comment]
Yes, I think this should be closed for Revise

So, Revise works in REPL, in the Juno REPL but not run from a julia file in Atom (julia 0.6.2)

I have added an issue at Juno:
https://github.com/JunoLab/Juno.jl/issues/107
Not sure whether the CC worked, but I put your handle in there
Also not sure whether it belongs where I put it (Juno) or atom-julia-client or Atom.jl or ....

Anyways, many thanks for writing Revise, much needed
